utils/logalloc.cc

```diff
@@ -846,80 +846,6 @@ class segment_store {
         }
     }
 };
-#if 0
-class segment_store {
-    std::vector<segment*> _segments;
-    std::unordered_map<segment*, size_t> _segment_indexes;
-    static constexpr size_t _std_memory_available = size_t(1) << 30; // emulate 1GB per shard
-    std::vector<segment*>::iterator find_empty() noexcept {
-        // segment 0 is a marker for no segment
-        return std::find(_segments.begin() + 1, _segments.end(), nullptr);
-    }
-    std::vector<segment*>::const_iterator find_empty() const noexcept {
-        // segment 0 is a marker for no segment
-        return std::find(_segments.cbegin() + 1, _segments.cend(), nullptr);
-    }
-
-public:
-    size_t non_lsa_reserve = 0;
-    segment_store() : _segments(max_segments()) {
-        _segment_indexes.reserve(max_segments());
-    }
-    const segment* segment_from_idx(size_t idx) const noexcept {
-        assert(idx < _segments.size());
-        return _segments[idx];
-    }
-    segment* segment_from_idx(size_t idx) noexcept {
-        assert(idx < _segments.size());
-        return _segments[idx];
-    }
-    size_t idx_from_segment(const segment* seg) const noexcept {
-        // segment 0 is a marker for no segment
-        auto i = _segment_indexes.find(const_cast<segment*>(seg));
-        if (i == _segment_indexes.end()) {
-            return 0;
-        }
-        return i->second;
-    }
-    std::pair<segment*, size_t> allocate_segment() noexcept {
-        auto p = aligned_alloc(segment::size, segment::size);
-        if (!p) {
-            return {nullptr, 0};
-        }
-        auto seg = new (p) segment;
-        poison(seg, sizeof(segment));
-        auto i = find_empty();
-        assert(i != _segments.end());
-        *i = seg;
-        size_t ret = i - _segments.begin();
-        _segment_indexes[seg] = ret;
-        return {seg, ret};
-    }
-    void free_segment(segment *seg) noexcept {
-        seg->~segment();
-        ::free(seg);
-        size_t i = idx_from_segment(seg);
-        assert(i != 0);
-        _segment_indexes.erase(seg);
-        _segments[i] = nullptr;
-    }
-    ~segment_store() {
-        for (segment *seg : _segments) {
-            if (seg) {
-                seg->~segment();
-                free(seg);
-            }
-        }
-    }
-    size_t max_segments() const noexcept {
-        return _std_memory_available / segment::size;
-    }
-    bool can_allocate_more_segments() const noexcept {
-        auto i = find_empty();
-        return i != _segments.end();
-    }
-};
-#endif
```

Comment:
This style of building a replacement in a different place and then deleting the original in a different patch results in small patches, but they're hard to review. How can I tell some detail wasn't lost?

Reply:
One of the goals of this series was to discard the vector-based `segment_store` and keep the `memory_layout`-based one only, with the differences in how that memory is acquired and managed moved into the new `segment_store_backend`.
The `memory_layout` based store implementation is modified in-place. We don't want to keep anything from this one.